### ISSUE TITLE
Use standard Canvas `ctx.filter` API (blur/grayscale/saturate/contrast)

### ISSUE THREAD

**Issue by paramaggarwal** [Automattic/node-canvas#1063]
HTML5 Canvas recently introduced support for filters like `ctx.filter = "blur(5px)";` in the browser, for example to `blur`, `saturate`, `grayscale` and modify `contrast`.

## Issue or Feature
API support for HTML5 Canvas `ctx.filter` from [from MDN web docs](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/filter)

## Steps to Reproduce
```js
var Canvas = require('canvas');
var canvas = Canvas.createCanvas(200, 200);
var ctx = canvas.getContext('2d');
ctx.filter = "blur(5px)";
ctx.drawImage(img,0,0);
```

## Your Environment
* node-canvas: 2.0.0 alpha
* node: v9.3.0
* macOS: High Sierra

**Comment by asturur** [comment]
This api is actually much new also for browsers ( chrome 49+ )

**Comment by paramaggarwal** [comment]
Ah, right. Closing issue for now.

**Comment by asturur** [comment]
I did not mean to make you close it. Just can be expected that is not implemented in  node-canvas since it became a standard non experimental api not long ago.

**Comment by paramaggarwal** [comment]
No, no - I previously thought that this is a bug, but actually it is a feature request. Hence closed it. 😄

**Comment by zbjornson** [comment]
Feature requests are valid issues :-)

**Comment by binh27112004** [comment]
Same problem. I make a canvas with nodejs and use ctx.filter = 'blur(4px)'; to blur canvas then I convert it to png. 
![image](https://user-images.githubusercontent.com/46067234/78055756-380b7480-73ae-11ea-8493-28180f824001.png)
It doesn't work !!
My code:
`
const canvas = createCanvas(900, 900)
const ctx = canvas.getContext('2d')
ctx.filter = 'blur(400px)';
ctx.font = '48px serif';
ctx.fillText('Hello world', 50, 100);
fs.writeFileSync('layer-1.png', canvas.toBuffer());
`

**Comment by asturur** [comment]
i would like to work on this, i wonder if we have the exact specs of the filters. Would be ok to implement those with normal loops over the pixels?

**Comment by zbjornson** [comment]
Reopening because we still don't have this feature.

@asturur I think the full list is here: https://drafts.fxtf.org/filter-effects/#supported-filter-functions. 

> Would be ok to implement those with normal loops over the pixels?

It's better to use Cairo APIs if possible because Cairo and pixman use SIMD extensions. However, the Cairo cookbook recipe for Gaussian blur (https://www.cairographics.org/cookbook/blur.c/) is indeed a loop. (Skia on the other hand has quite a few filters built in, were we to switch to that.)

**Comment by paramaggarwal** [comment]
Oh ok, I was going through all my open issues across GitHub and closing old ones. Thanks for reopening.

**Comment by asturur** [comment]
I have a dumb question, since so long and i never made it just because i was worried it was too dumb.

To get a great node-canvas compatibility, what does stop someone to take the full firefox canvas implementation from the source code and make it for node?

Wouldn't it an accurate node-canvas at that point?

**Comment by GodyFromDiscord** [comment]
Can the Cairo API still be used with node canvas to blur then carry on using canvas as normal for the rest?

**Comment by mifi** [comment]
Hm, I found this: https://github.com/Automattic/node-canvas/blob/master/src/CanvasRenderingContext2d.cc#L615
Looks like an implementation of the filters 😮

**Comment by HoseinDanesh** [comment]
issue in 2017 
awful
Doesn't want provide this feature ?
whatever its experimental or not

**Comment by Saberjsd** [comment]
I had the same problem........

**Comment by PainOchoco** [comment]
Any update on that? It would be really cool!

**Comment by PainOchoco** [comment]
In the meantime I will use [skia-canvas](https://github.com/samizdatco/skia-canvas) which supports filters, it's a great alternative.

**Comment by dawidjaniga** [comment]
Hi, do you plan to implement `filter`?

**Comment by zmnv** [comment]
we really need a blur filter please < 3

**Comment by paramaggarwal** [comment]
To support filters we will need to migrate to Skia graphics engine from the Cairo engine. Such a rewrite would effectively be like starting an entirely new project and should not be in the scope of this project (which should continue to be a great canvas API wrapper on top of Cairo engine) - hence closing this.

See discussion here: https://github.com/Automattic/node-canvas/issues/1652#issuecomment-691535373

**Comment by Pomax** [comment]
Another option is to implement filters "the old way" (i.e. composite to a new pixel matrix, then replace the original with the updated one) and tell folks that this will work, but will be slow (but for most applications that would probably be an acceptable tradeoff. I'd rather wait a little longer on an image generation if it means the result looks identical between node-canvas and the browser =)
